protocol 1.7: outpoint subs

SomberNight opened the pull request:
This defines a new Electrum Protocol version: 1.7.

changes compared to 1.6:
- the `blockchain.scripthash.*` methods are all replaced with `blockchain.scriptpubkey.*` methods
  - ref https://github.com/spesmilo/electrum-protocol/issues/3
  - ~~the new methods also contain the `chaintip` in their response~~
    - reverted in https://github.com/spesmilo/electrum-protocol/pull/17
- new RPCs: `blockchain.outpoint.subscribe` RPC (and related RPCs) to subscribe to a transaction
  outpoint, and get a notification when it gets spent.
- `server.ping` can now also be sent as an unrequested notification, by both the client
  and the server. The request/response signature also changed.
- standardize "history too large" error code for `blockchain.scriptpubkey.*` methods
- new RPC: `blockchain.transaction.testmempoolaccept`
  - ref https://github.com/spesmilo/electrum-protocol/issues/11 
- new RPC: `mempool.recent` to list a few of the latest txs just added to the mempool.
  - ref https://github.com/spesmilo/electrum-protocol/issues/13

---

This proposal is less ambitious than previous drafts of 1.7:
- most notably witness SPV is postponed
  - ref https://github.com/spesmilo/electrum-protocol/pull/16
- the `method_flavours` field for the `server.features` response is also postponed,
  as I realised we are not clear re how to signal/negotiate optional features
  - see related https://github.com/spesmilo/electrum-protocol/issues/7 idea
- `blockchain.outpoint.subscribe` requires `spk_hint` as an input param, to help server impls

---

We have a client implementation in https://github.com/spesmilo/electrum/pull/10627,
and a server implementation in https://github.com/spesmilo/electrumx/pull/303.

---

EDIT: note that https://github.com/spesmilo/electrum-protocol/pull/17 was an after-merge revert, which reverted adding the chaintip to RPC responses

SomberNight committed:
add new protocol version 1.7

- This is mostly split off from the protocol 1.5/2.0 specs and ideas
  in https://github.com/spesmilo/electrumx/pull/90
  - notably it does not include the history pagination, and the recursive scripthash status,
    which are deferred for now

Armanidashh reviewed (commented):
1

SomberNight commented on docs/protocol-methods.rst:
```diff
@@ -546,6 +752,72 @@ Protocol version 1.0 returning an error as the result:
 
   "258: txn-mempool-conflict"
 
+blockchain.transaction.broadcast_package
+========================================
```
note: this is from https://github.com/spesmilo/electrum-protocol/pull/1

f321x replied:
I made some changes in this commit https://github.com/spesmilo/electrum-protocol/pull/1/commits/7bc5f11efc6f0b39a2958ea2ccac63aa763fe1f6 that add an additional, optional errors field to the non-verbose message, replace the package_msg with a success: bool and make the replaced txids field optional too.

SomberNight replied:
changes incorporated

SomberNight replied:
needs updates from https://github.com/spesmilo/electrum-protocol/pull/1 again

SomberNight replied:
done again

SomberNight committed:
add "method_flavours" to "server.features"

SomberNight committed:
list more protocol changes

SomberNight committed:
re-design "server.ping", to be used against traffic analysis

SomberNight committed:
re-design "server.ping", to be used against traffic analysis

SomberNight committed:
add 'block_hash' field to 'transaction.get_merkle'

related: https://github.com/kyuupichan/electrumx/issues/385

SomberNight committed:
use scriptPubKeys everywhere instead of scripthashes

related https://github.com/spesmilo/electrum-protocol/issues/3

SomberNight committed:
use scriptPubKeys everywhere instead of scripthashes

related https://github.com/spesmilo/electrum-protocol/issues/3

SomberNight committed:
(TODO) witness-spv

SomberNight committed:
add section: buffer and pad msgs to protect against traffic analysis

SomberNight committed:
add section: buffer and pad msgs to protect against traffic analysis

SomberNight committed:
add new RPC: "mempool.recent"

SomberNight commented on docs/protocol-methods.rst:
```diff
@@ -467,50 +498,243 @@ Return an ordered list of UTXOs sent to a script hash.
+
+**Signature**
+
+  .. function:: blockchain.outpoint.subscribe(tx_hash, txout_idx, spk_hint=None)
```
note: bitcoin core added a `txospenderindex`, to be used with the `gettxspendingprevout` RPC.
ref https://github.com/bitcoin/bitcoin/pull/24539

- see what their API looks like and maybe adapt what we add to that
  - in particular we could put `spender_blockhash` in the response
  - they also have an option to put the full tx in the response

SomberNight replied:
note: electrumx server impl was changed to use `txospenderindex` instead of a custom re-implementation of essentially the same.
Re the protocol API, we did not change it.

SomberNight committed:
outpoint.subscribe: flatten notification params

This construction makes it more homogeneous with existing notifications.

ref https://github.com/spesmilo/electrumx/pull/303#discussion_r3167100679

SomberNight committed:
add new RPC: testmempoolaccept

SomberNight committed:
add new RPC: testmempoolaccept

ecdsa commented on docs/protocol-methods.rst:
```diff
@@ -1072,6 +1543,18 @@ Return a list of features and services supported by the server.
+    If a server supports all functionality defined for the negotiated protocol version,
+    it can just set this to the empty dict (but the `method_flavours` key itself
+    must always be present).
+
```
I think `method_flavours` should be postponed for now, because its role is not clear.
it seems that what it wants to achieve collides with the features negotiation proposal in #7

SomberNight replied:
OTOH it would be good if the server could clearly signal whether it requires `spk_hint` for `blockchain.outpoint.subscribe`.

`method_flavours["blockchain.outpoint.subscribe"]["requires_spk_hint"]=true` was the motivating use case for `method_flavours`. (see e.g. https://github.com/spesmilo/electrum-protocol/commit/4b4dcdacc2c4d703d42b72505e0fae9d6f29eb1a)

Or do we accept that there is no way for a client to know whether the server requires spk_hint, for now?

SomberNight replied:
We could make `spk_hint` a mandatory field for now, and perhaps make it optional in a future version with proper feature flags. In all reasonable scenarios, the client should know the corresponding spk.

SomberNight replied:
> We could make `spk_hint` a mandatory field for now

done in https://github.com/spesmilo/electrum-protocol/commit/4460b8564ea29e49136f05618e7ee075f738a142

SomberNight committed:
standardize "history too large" error code

SomberNight committed:
add new RPCs: "blockchain.outpoint.subscribe" and friends

SomberNight committed:
RPC: "blockchain.outpoint.subscribe": make spk_hint mandatory for now

f321x commented on docs/protocol-basics.rst:
The example is still using scripthash methods

SomberNight replied:
changed in https://github.com/spesmilo/electrum-protocol/compare/acc7d161188c62efba6838748e26dcc17e8193b3..e8a379c369cec7ba1edb34951044243d9cfa5c79

f321x commented on docs/protocol-changes.rst:
```diff
@@ -228,11 +228,29 @@ Changes
 
   * :func:`server.ping` can now also be sent as an unrequested notification, by both the client
     and the server. The request/response signature also changed.
+  * The `blockchain.scripthash.*` methods are all replaced with `blockchain.scriptpubkey.*`
```
Should the removal of `hash_function` from `server.features()` be mentioned as change too?

SomberNight replied:
sure.
changed in https://github.com/spesmilo/electrum-protocol/compare/acc7d161188c62efba6838748e26dcc17e8193b3..e8a379c369cec7ba1edb34951044243d9cfa5c79

SomberNight committed:
readthedocs: update version str

f321x commented on docs/protocol-methods.rst:
```diff
@@ -316,6 +316,10 @@ Return the confirmed and unconfirmed history of a :ref:`scriptPubKey <scriptpubk
   transactions are returned.
 
+  If the history of the scriptPubKey is too long (busy address, touched by thousands of txs),
+  and so the server refuses to serve it, it SHOULD send a JSON-RPC error with integer code `10_001`
```
For blockchain.transaction.broadcast_package we use the error code 1, maybe it would be useful to collect them in a table somewhere?

Also it might be safer to write 10001 instead of 10_001 to prevent implementers from trying to incorrectly send/parse the literal 10_001 string instead of the int.

SomberNight replied:
> For blockchain.transaction.broadcast_package we use the error code 1,

That's just the error code used in the example traffic.
ElectrumX uses `BAD_REQUEST = 1`, `DAEMON_ERROR = 2`, but even that is not used consistently, and it is specific to ElectrumX.

> maybe it would be useful to collect them in a table somewhere?

Yes, that would make sense, I agree. Except this is the first error code that we do specify, so maybe it is too early for that.

> Also it might be safer to write 10001 instead of 10_001 to prevent implementers from trying to incorrectly send/parse the literal 10_001 string instead of the int.

I think that syntax is used in multiple programming languages, so it should be understandable, but ok, I will change it anyway.

SomberNight replied:
> safer to write 10001 instead of 10_001

changed in https://github.com/spesmilo/electrum-protocol/compare/acc7d161188c62efba6838748e26dcc17e8193b3..e8a379c369cec7ba1edb34951044243d9cfa5c79

f321x commented on docs/protocol-changes.rst:
```diff
@@ -228,11 +228,29 @@ Changes
+Removed methods
+---------------
+
+  * :func:`blockchain.scripthash.get_balance`
```
Previously removed methods (e.g. `blockchain.address.*`) were added to `protocol-removed.rst`. 
https://github.com/spesmilo/electrum-protocol/blob/master/docs/protocol-removed.rst
Should the `blockchain.scripthash.*` family go in there as well?

Also note the `blockchain.address.*` methods mention stale references to the `blockchain.scripthash.*` methods.

SomberNight replied:
> removed methods [...] added to `protocol-removed.rst`

Oh right! I forgot about that.
Added in https://github.com/spesmilo/electrum-protocol/compare/acc7d161188c62efba6838748e26dcc17e8193b3..e8a379c369cec7ba1edb34951044243d9cfa5c79.

> Also note the `blockchain.address.*` methods mention stale references to the `blockchain.scripthash.*` methods.

I think that's fine now that the `bc.sh.*` methods are also there in `protocol-removed.rst`.

ecdsa commented on docs/protocol-methods.rst:
```diff
@@ -477,6 +481,10 @@ Subscribe to a :ref:`scriptPubKey <scriptpubkeys>`.
 
+  If the history of the scriptPubKey is too long (busy address, touched by thousands of txs),
+  and so the server refuses to serve it, it SHOULD send a JSON-RPC error with integer code `10_001`
+  to communicate this to the client.
```
Edge case: if the history is not too long when you subscribe, but then it becomes too long. 
Should the server send a notification in that case?
Error codes in jsonrpc only exist in responses, not in notifications.

SomberNight replied:
>  Error codes in jsonrpc only exist in responses, not in notifications.

Right, I also looked this up exactly for this reason a couple days ago :D

Well the way ElectrumX has always worked in this case is that the server silently drops the subscription at that point. Not ideal but I would leave that as-is.

The protocol 2.0 history pagination and incremental recursive status would be the true fix.

ecdsa replied:
We could return zeroes as the hash in the notification.
The client will request history if he receives a hash that does not match its own.

> The protocol 2.0 history pagination and incremental recursive status would be the true fix.

indeed

SomberNight replied:
> Well the way ElectrumX has always worked in this case is that the server silently drops the subscription at that point. Not ideal but I would leave that as-is.

I was wrong. ElectrumX drops the subscription but also sends back a notification with a null status (as if the scripthash did not have any history).
That should prompt the client to call get_history at which point they will get the JSON-RPC error.

I don't think this needs to be spec'd out.

f321x commented on docs/protocol-methods.rst:
```diff
@@ -1308,18 +1308,112 @@ subscription and the server must send no notifications.
+
+  * *data*
+
+    A hexadecimal string (but can be odd-length). Its value is to be ignored by the recipient.
```
Maybe `A string of hexadecimal characters of arbitrary length (including odd lengths).` makes more clear that this string doesn't have to be a valid hex encoded bytes?

SomberNight replied:
Hmm. For the response the length is not arbitrary, so that phrasing is not perfect either.
There are both odd-length and even-length examples a few lines below. I think it's fine as-is.

SomberNight committed:
traffic analysis: say "SHOULD" for padding, "COULD" for timing delays

SomberNight commented on docs/protocol-basics.rst:
```diff
@@ -256,3 +262,123 @@ and confirm the returned roots match.
+
+- pad messages to bucketed lengths (e.g. powers of 2, with a min size), and
+
+- introduce small timing delays, ideally by buffering messages.
```
@craigraw

> one major concern: The new `Traffic analysis` section bundles two different things under one SHOULD:
> - **Padding**: reasonably cheap and backwards-compatible, needs no handshake. Fine as a SHOULD in 1.7.
> - **Buffering / timing delay**: costs interactive latency and the client can't refuse it. Should be COULD, or gated behind feature negotiation.
>
> [...]
>  My recommendation would be to hold the timing delay recommendation, or downgrade it to COULD, until #7 lands the response-side `options` field and a `delays` feature definition.
> [...]
> Either way, I believe the recommendation of timing delays must be balanced with the ability to control them from the client side.

Changed in 1a58f6caf98fae03d36bc6570190e936394f8e08 to explicitly say SHOULD for padding and COULD for timing delays.

I agree that it would be desirable for the client to be able to influence the server's behaviour here via an opt-out, e.g. by negotiating a max timing delay. Still, I would postpone that for later.

romanz reviewed (commented):
LGTM, thanks!

t-bast reviewed (commented):
It looks good from my point of view as well. I'll let @sstone take another look to make sure this won't create issues for Phoenix.

tnull commented on docs/protocol-basics.rst:
```diff
@@ -256,3 +262,123 @@ and confirm the returned roots match.
+
+As a generic mitigation, implementations (both client and server)
+
+- SHOULD pad messages to bucketed lengths (e.g. powers of 2, with a min size), and
```
Very much appreciate the push to preserve more privacy!

Note that padding on the transport layer could be more efficient. I.e. rather than padding each individual message, implementations could make sure they send out TCP frames of uniform lengths. Then, the only way missing would simply be a new variable-length padding message types (e.g. ping with a variable length payload) that allows to fill up remaining space whenever a message is to be sent that doesn't exactly fit into a multiple of the designated length value. We are now implementing this approach in Lightning land.

SomberNight replied:
> rather than padding each individual message, implementations could make sure they send out TCP frames of uniform lengths

The protocol supports multiple underlying transports: TCP, SSL, WS, WSS.
(The "TCP" and the "WS" transports are not encrypted so out-of-scope here.)
The most popular transport is probably SSL.

How much control does a userspace application have over the TCP frame when using e.g. the OpenSSL library?
At least when interacting with OpenSSL from the python stdlib, in our case, we don't have a lot of control.
What we have control over is what we feed into the SSL stream.

In the websocket case, it's probably even less control.

> We are now implementing this approach in Lightning land.

We currently use SSL instead of BOLT-08 :(
With BOLT-08, indeed you have ~direct control of the TCP stream, but the situation seems to be more complicated here.

tnull replied:
> How much control does a userspace application have over the TCP frame when using e.g. the OpenSSL library?
> (...)
> In the websocket case, it's probably even less control.

Hmm, granted. I think you could always implement your own buffered message queue, disable nagle, and at least push out constant-sized chunks. But yea, probably you have little control over what happens after in another layer.

tnull commented on docs/protocol-basics.rst:
```diff
@@ -256,3 +262,123 @@ and confirm the returned roots match.
+
+- COULD introduce small timing delays, ideally by buffering messages.
+
+(A future protocol version will allow the client to opt-in/opt-out of this server-behaviour,
```
Tbh, this is a bit odd. If we want to make it optional going forward, maybe padding shouldn't be part of the spec until everything is ready?

SomberNight replied:
> Tbh, this is a bit odd. If we want to make it optional going forward, maybe padding shouldn't be part of the spec until everything is ready?

Fair enough, I see your point.

However, in this version, we want to extend "server.ping" so that it can be used to mimic traffic patterns in a generic way. That's the only thing that is new in 1.7.

All the rest of the traffic analysis text is basically independent of protocol versions and is just describing what Electrum Wallet and ElectrumX have already been doing for the past year or so. It describes one specific way of doing things and gives some recommendations. It is bundled as part of this PR because I felt that rationale was needed for why "server.ping" is being extended.

The opt-in/opt-out would be a new thing that's not there yet. It is mentioned as it would be useful and is somewhat important, however I would not hold up the rest of the changes for it. We need to do smallish incremental changes as otherwise - experience suggests - all progress is stalled.

tnull commented on docs/protocol-basics.rst:
```diff
@@ -256,3 +262,123 @@ and confirm the returned roots match.
+they send and not the server (or the other way around),
+that just limits the effectiveness of the defense against traffic analysis.
+
+Note when the JSON-RPC messages are sent in the TLS stream, they are sometimes batched together.
```
Right, but why not make this the default way of doing padding? The more ambiguity the higher the chances will be that implementations will mess up or specific clients will stand out, i.e., still end up leaking data to observers.

SomberNight replied:
> Right, but why not make this the default way of doing padding?

Do you mean that instead of injecting whitespaces, why not send json-rpc "Batches" with one of the requests in the batch being a "server.ping" notification with a custom data length? Yes that works too. (note that the "server.ping" notification is a new addition in protocol 1.7. In older protocol versions there is ~nothing the server could send.)

It is trivial to implement padding by injecting extra whitespaces. It also has high granularity: you can add any number of bytes of padding, even just one byte.

Whitespace-injection can be done at any protocol version as json parsers just ignore it. It can even be done at a lower level, abstracted away from the electrum session logic. I like this approach the most.

Conceptually even "batching" could be done in two different ways: using the JSON-RPC "Batch" concept, or just using a buffer and sending out multiple JSON-RPC messages at the same time. The main difference is that using the JSON-RPC "Batch" concept forces the responder to similarly batch the responses.

So, yes you are right there is a high degree of freedom.

In any approach one might want to do timing delays because it allows adding more padding by amortising the cost of the padding bytes over more payload data. So batching (in whatever form) needs to be mentioned here.

> The more ambiguity the higher the chances will be that implementations will mess up or specific clients will stand out, i.e., still end up leaking data to observers.

I think it's probably not realistic to make sessions indistinguishable across implementations. Already the choice of server a client communicates with typically gives a good clue regarding which client implementation is being used. The choice of server can be inferred either based on the IP address or from the hostname in the TLS Client Hello. If the sniffing is done close to the client, the number of simultaneous server connections is another fingerprinting vector: if it's more than one, it's probably Electrum Wallet.

tnull replied:
> Do you mean that instead of injecting whitespaces, why not send json-rpc "Batches" with one of the requests in the batch being a "server.ping" notification with a custom data length? Yes that works too.

Yes. Reading the current version I felt that it describes too many ways to achieve the same thing, which usually either means something is underspecified or the given examples can be reduced/reformulated in an abstract way. Adding too much verbosity usually just leads to confusion and devs either messing up or not doing anything at all. 

So, not necessarily saying that batching is a waaay preferable approach, but it could be preferable to reduce the guidance to one suggested/preferred way of introducing padding. If you think whitespace padding is the preferable approach, well, then maybe that should be the only approach described, if only to keep the spec leaner?

tnull commented on docs/protocol-methods.rst:
```diff
@@ -464,50 +463,251 @@ Return an ordered list of UTXOs sent to a script hash.
+  Similarly, the server MAY ignore new subscription requests if the spending tx is already
+  mined at a reorg-safe height but it still MUST send at least one full response.
+
+**Result**
```
It might be useful if this (as well as similar APIs like `get_balance`, etc) would include a field with the current tip hash. When we make multiple subsequent requests it's otherwise always hard to tell whether the chain moved while we're at it.

SomberNight replied:
> similar APIs like get_balance, etc

I guess you have these methods in mind:

- `blockchain.scriptpubkey.get_balance`         - dict
    - I guess tip could make sense
    - though this method cannot be SPV-ed, so it is fully trusting the server. Not sure how widely it is used in the wild.
- `blockchain.scriptpubkey.get_history`         - list => cannot easily add tip
    - should be used with sub??
- `blockchain.scriptpubkey.get_mempool`         - list => cannot easily add tip
    - not sure how people use this method...?
- `blockchain.scriptpubkey.listunspent`         - list => cannot easily add tip
    - I guess tip could make sense
- `blockchain.outpoint.get_status`              - dict
    - I guess tip could make sense
- `blockchain.outpoint.subscribe`               - dict
    - if tip raced, client will just get a new notification
    - note btw current text says (so would have to be tweaked a bit):
        > "any event that changes any field of the `status` dictionary results in a notification"

For both of `bc.outpoint.get_status` and `bc.outpoint.subscribe`, if you then subsequently SPV the txids, that mostly solves the issue.

Though if you called `bc.outpoint.get_status` just when the tip also moved, and a relevant tx just moved from mempool to the new block, then you would not know you received stale data (still mentioning the mempool) - but indeed you would if we included the chaintip in the response.

In our use case, we use the subscriptions, and SPV the txid, so there is no ambiguity. What kind of client-usage pattern do you have in mind?

tnull replied:
Well, eventually we may wan to switch to make use of subscriptions (and with outpoint subscriptions that could finally work out, we'll def. try once this is shipped and has sufficient support) but for now we mostly retrieve all data via polling (cf https://github.com/lightningdevkit/rust-lightning/blob/main/lightning-transaction-sync/src/electrum.rs). So having the tip hash available as relevant metadata for reorg detection would be great in general, especially for all methods that are not subscription based, e.g., `get_status` etc. IMO any HTTP based protocols should just include it as a reponse header, in Electrum it would probably mean adding it as an optional field ~everywhere if possible, though that could of course also happen over time.

SomberNight replied:
@LLFourn had similar remarks out-of-band yesterday.

Given that some of the RPCs return lists, while others return dicts, but adding the chaintip is problematic to lists,
and given that as part of the scripthash->spk change we are effectively adding new methods anyway,
I guess now is the time to "change" ~all the return types to dicts.

So, we could include the chaintip in the responses.

I wrote a spec for it now, though I have not written client/server code yet. The server might not be trivial to write, as it has to check for reorgs and retry, but it looks doable.

Please see if https://github.com/SomberNight/electrum-protocol/commit/9c32658b4629196f8ff7043c210beb0786bc0dd1 looks good.
I have not pushed that commit to this branch yet, given that ultimately it is you who knows what exactly you want, but if it looks good, I would push it here and maybe we could still include it in 1.7.

tnull replied:
Ah, cool, thanks! Haven't reviewed the commit fully in close detail yet, but concept ACK! I'm not quite sure if the block hash compression is really necessary, but you are likely more knowledgeable regarding what data/indexes are already held in the implementations, so I'll refrain from opining on that. Agree 1.7 could  be the right time to make the change.

SomberNight replied:
I wrote [server-side](https://github.com/spesmilo/electrumx/pull/362) and [client-side](https://github.com/spesmilo/electrum/pull/10627) code for this, and it looks fine.
The client is of course trivial, the server-side was more interesting.

I have now pushed [that spec commit](https://github.com/SomberNight/electrum-protocol/commit/9c32658b4629196f8ff7043c210beb0786bc0dd1) to this branch.

SomberNight replied:
Could @romanz, @cculianu, or someone working on a server, comment if this seems reasonable to add into their server code?

-----

In general it is not trivial for a server to add.
To illustrate why, I will describe briefly how it looks like in ElectrumX:

The server is not monolithic. bitcoind and electrumx are two separate processes (not necessarily running on the same machine). e-x maintains its own mempool, so the mempool is essentially duplicated.

As far as e-x is concerned, bitcoind has "two" heights: daemon.height() calls a bitcoind RPC directly to ask for the current height, while daemon.cached_height() just internally returns the last height seen returned from daemon.height().
If e-x realises it's behind, it will fetch blocks and process those. Once they are processed in-memory, BlockProcessor.height is bumped. Then e-x flushes the db changes to the disk, after which db.height is bumped.
Meanwhile e-x's mempool logic is periodically polling bitcoind using getrawmempool, but only if daemon.height() == db.height.
Finally, after a new block, notifications are only sent to clients once a mempool polling tick finished, and then notified_height is bumped.

So there are ~5 different heights:
- `daemon.height()`
- `daemon.cached_height()`
- `block_processor.height`
- `db.height`
- `notified_height`

The propagation is in that order: daemon.height() -> daemon.cached_height() -> bp_height -> db_height -> notified_height

I am still thinking about potential designs re how to give this consistency guarantee to the client, but the following is what I have right now.

It's probably not practical to take a lock when any of thousands of clients calls a db query method to read or when the db is being updated. So what I implemented in e-x is a hack: when e.g. a get_history RPC is being handled, just before and just after the server can check the current height. When getting the height, it ensures:
`daemon.cached_height() == block_processor.height == db.height == notified_height`, and that's the value it uses.
If the height is the same when starting to process the request and when finishing it, then it will put that chaintip into the response and send it. Otherwise, it retries a few times. Specifically before retrying it sleeps until notified_height changes (with a timeout). If the retry attempts are exhausted, it sends a json-rpc-error to the client.

~~I believe it's mostly sufficient to compare heights and not full tip hashes, as a reorg between chains of the same length are highly unlikely (but I guess could happen). (maybe should compare tip hashes...)~~ (EDIT: comparing tip hashes now, and also comparing "reorg count")

On my mainnet server it takes 1-5 seconds to process a block (typically 2s). I saw one block that took 10 seconds to process. This is on fast modern hardware, limited by python single-core CPU usage and disk I/O.
While the BlockProcessor is processing a block, relevant RPC calls would be blocked.

Previously RPCs would be processed concurrently to all this, operating on the DB and e-x's mempool.
The DB would always be self-consistent and the mempool would always be self-consistent, but actually the mempool might be at an older height than the DB. (it could e.g. happen that a get_history result contains a just-mined tx twice: both in a block and in the mempool -- this is clearly a bug, though due to other internal implementation details it is even less likely to happen)

EDIT: realised it is ok if bitcoind is at a different height, just need to make sure e-x itself is in a consistent state (`block_processor.height == db.height == notified_height`). What can take several seconds is processing a block, but while that is being done, the server can just serve the old state from before the block. This limits the critical time window to <2 sec.

-----

If the protocol wants to add this consistency guarantee, that shifts the burden of figuring out state from the client to the server.  Clearly that's highly desirable to some clients, so the question is how practical it is for a server to implement.

Also, as a client, would you rather wait a few more seconds to get back a guaranteed-consistent response (in case a block is racing), or would you just want any response but fast? (EDIT: I believe I managed to make e-x fast enough so that the tradeoff is obviously worth it)

SomberNight replied:
@mononaut I see there is related discussion in https://github.com/mempool/mempool/issues/6584 but I don't have permissions to comment there. As we are considering adding the chain tip to json-rpc responses in the electrum protocol now (as seen in current PR changeset), it would be good to know if that is something some server implementations think they cannot reasonably support.

SomberNight replied:
> realised it is ok if bitcoind is at a different height, just need to make sure e-x itself is in a consistent state (`block_processor.height == db.height == notified_height`). What can take several seconds is processing a block, but while that is being done, the server can just serve the old state from before the block. This limits the critical time window to <2 sec.

(The critical window is the time window when if a protocol request comes in, serving it will be delayed until after the critical window ends.)

I have now made the mempool processing code in e-x faster, changed it to be event-based instead of polling, and made it run partially concurrent to the new-block-processing. Now the critical time window in e-x is <1 sec, typically around 0.4 sec, on a busy server.

This performance I believe is perfectly acceptable, and shows it is practical to tie back protocol responses to specific chaintips. (without rocksdb snapshots or similar, and while having a separate mempool state and chainstate)

Also, @romanz has confirmed out-of-band that adding the chaintip to protocol responses and ensuring the protocol response is consistent with said chaintip should not be too hard in electrs.

tnull reviewed (commented):
Thanks, mostly LGTM

Just a few comments regarding the padding. IMO it could make sense to just do it at the TCP level rather than for each individual message. And in general it might be good to prescribe *one* way to do it rather than leaving too many degrees of freedom, as all of them increase confusion and the risk of implementers end up doing something that makes them leak information to an observer.

ecdsa commented on docs/protocol-methods.rst:
```diff
@@ -464,50 +463,251 @@ Return an ordered list of UTXOs sent to a script hash.
+    .. note::  Full index servers might not need the parameter in practice
+      but some lighter personal servers (such as EPS, BWT, Floresta)
+      would not be able to serve the request without it. A future protocol version might
+      add a feature flag where the server can signal whether it requires spk_hint.
```
I would remove that sentence (regarding future protocol version), it is not useful here, and it is not clear whether we want to do that.

SomberNight replied:
done in https://github.com/spesmilo/electrum-protocol/commit/a3f65010dca693cb51b296455c94996eb74c636c

SomberNight committed:
address feedback: rm mention of future version spk_hint opt, grammar

(feedback from ThomasV)

Davidson-Souza commented:
Concept ACK; thank you for adding the `spk_hints` to `blockchain.outpoint.subscribe` as well.

SomberNight commented on docs/protocol-changes.rst:
```diff
@@ -216,3 +216,45 @@ Removed methods
+  * :func:`server.ping` can now also be sent as an unrequested notification, by both the client
+    and the server. The request/response signature also changed.
+  * The `blockchain.scripthash.*` methods are all replaced with `blockchain.scriptpubkey.*`
+    methods. Note: `sha256(scriptPubKey) == scripthash`.
```
@cculianu   
  
> - removing `blockchain.scripthash.*` and replacing them with `blockchain.scriptpubkey.*` is a mistake. The scripthash thing is clever for a variety of reasons. Support both. Internally the spk one can just map to scripthash impl. anyway.

For a full index server it is indeed trivial and cheap to support both. As the notifications still use scripthashes, the server would need to keep some scripthash logic, also the existing servers already have DBs that reference scripthashes. So yes, the idea is that a server impl would internally map spks to scripthashes anyway.

> makes no sense to remove them it’s actually superior for many use cases

When are they superior?
I see the following advantages for scripthashes, all of them weak:
1. they are constant-length, which is good against traffic analysis
    - but then you also said for server.ping:
       > I realize this is for "privacy" but I think it's a waste of time to even worry about. 
    - instead we should just pad the traffic, as also described in this changeset
2. they use less bandwidth compared to long scriptpubkeys
    - but note that typical popular scriptpubkey types are of similar length:
      (from top of my head, ignoring hex, p2wpkh is 22 bytes, p2wsh is 34 bytes, p2tr 34 bytes, vs scripthashes are 32 bytes)
3. I remember kyuupichan's argument in the past was that they are good for privacy against low-skill would-be-chainalysis server operators, as if the client directly sent over their addresses, the server op would have all their addresses linked together already
    - in contrast, if the client only sends a list of scripthashes, many of which are unused, the server op needs to store them in a database for a long time and only slowly over time would the addresses get revealed when they get used
    - also if a server was blanket saving all sets of subbed unused scripthashes together, they could be spammed (tricky clients could intentionally bloat their DB)
    - while an interesting thought experiment, against a high-tech adversary such as the actual Chainalysis and the like, I believe the argument to be moot

These were the arguments we already went through internally and disregarded them. Though indeed I had not mentioned them publicly in this issue before.

@evoskuil 

> Given that this is a proposed change, how could such servers been operational to date? And out of curiosity, what makes it challenging to provide an index of the hash?

See discussion linked from the OP:

> - the `blockchain.scripthash.*` methods are all replaced with `blockchain.scriptpubkey.*` methods
>   - ref https://github.com/spesmilo/electrum-protocol/issues/3

Basically light personal servers such as [EPS](https://github.com/chris-belcher/electrum-personal-server), [bwt](https://github.com/bwt-dev/bwt-electrum-plugin), and [Floresta](https://github.com/getfloresta/Floresta) can be used if you already have your own bitcoind but want to use an electrum-protocol-based wallet, and connect your wallet somehow to your bitcoind.
The way they work is that the user needs to configure the middleware with their xpub (or output script descriptors), and then
- either the middleware RPC-calls into bitcoind wallet to create a wallet using a list of imported addresses, and it is the middleware that pre-calculates the addresses and rolls ahead the gap limit
- or the middleware RPC-calls into bitcoind wallet (usually Bitcoin Core) to create a wallet using the output script descriptor
- or Floresta uses BIP-158 neutrino-style compact block filters to scan for the addresses, and those filters are indexed based on scriptpubkeys

(note that EPS and bwt are unmaintained but Floresta is very much alive)

The issue with keeping support for scripthashes is that then basically no client will migrate and Floresta and similar approaches will not be able to take advantage of the scriptpubkey subs in practice.

---

on another note, please try to keep top-level comments to a minimum: organise topics into threads, to keep it easier to follow.

ecdsa replied:
note: there is an in-development EPS plugin: https://github.com/svanstaa/electrum-eps-plugin

cculianu replied:
> These were the arguments we already went through internally and disregarded them. Though indeed I had not mentioned them publicly in this issue before.

Yes, you tend to do that. Just unilaterally make decisions. 

There is no *harm* in continuing to support scripthashes. Aside from the advantages you mentioned, here is another that you neglect:

- Why impose a cost on client software to update to support scriptPubKeys if it already has code written to think/talk in terms of scriptHashes?

I know it's a small update (after all to get to a scriptHash you need an spk first) -- but it's an update nonetheless!  Why impose this cost at all?  You can easily support both and you should.


> The issue with keeping support for scripthashes is that then basically no client will migrate and Floresta and similar approaches will not be able to take advantage of the scriptpubkey subs in practice.

Huh? I said support both.

SomberNight replied:
>> The issue with keeping support for scripthashes is that then basically no client will migrate and Floresta and similar approaches will not be able to take advantage of the scriptpubkey subs in practice.
>
> Huh? I said support both.

Floresta is a server. I am arguing that by supporting both, clients will never migrate. This is also what you are saying by the way, you are arguing clients should not need to be changed. Except then servers like Floresta will never get a chance to be adopted, because they will never work with lazy clients.

(by lazy client I mean clients that implement the new protocol but only make the most minimal changes so e.g. would not change from sh to spk)

Clients that don't implement changes will have to keep using old protocol versions.
That provides a clear incentive structure: if a client wants to take advantage of new features, they also have to adapt their existing code for other changes.

Years from now, servers might not support old protocol versions anymore, and at that point, old clients that still will not have updated, will get broken. I think that is fine. We don't want to or need to keep accumulating technical debt.

>> These were the arguments we already went through internally and disregarded them. Though indeed I had not mentioned them publicly in this issue before.
>
> Yes, you tend to do that. Just unilaterally make decisions.

As soon as someone asked, I listed the arguments.

So will you list the "actually superior for many use cases" use cases you had in mind? :)
"weak advantage" does not mean "superior"

The argument that we should keep supporting scripthashes because then clients do not have to be changed is extremely weak. By using a new protocol version, the client impl already needs to make a change! They need to signal they support it. Might as well implement some changes then. :)  Or don't change anything in the client impl and keep negotiating the existing protocol versions.

Technical feedback is very welcome, that's why I tagged people. I had also tagged you. Subjective opinions are also welcome. The tradeoffs need to be weighed, but ultimately someone needs to make decisions, or there will never be any progress.

This is not like Bitcoin consensus, forking away is much cheaper. We don't need a full democracy and don't need to argue over soft forks for 5 years. Unlike in consensus rules, we can also undo changes later if they turned out to be bad decisions. I don't want the protocol to ossify, I want to be able to change things. To deprecate and remove stuff. To try new things. For example, when scripthash subs were added, soon after address subs were removed. I think that was a good change, very much worth it. I don't want the server to have to know about address encoding/decoding at all. But AFAIU you still support address subs in Fulcrum. I strongly think it's not worth it to keep supporting that.

In my eyes, one of the main points of having version negotiation is to be able to remove functionality.

In general your feedback is appreciated on these matters, as you have implemented both server and client code using the protocol, so clearly you are a subject matter expert.

SomberNight committed:
add "chaintip" blockref field to most responses

also mandate notifications to be sent for subs even if visible height fields did not change but there was a reorg changing the "invisible" corresponding blockhashes

sstone commented on docs/protocol-methods.rst:
```diff
@@ -447,67 +490,292 @@ Return an ordered list of UTXOs sent to a script hash.
+  where the spending transaction is already deeply mined at a reorg-safe height (typically
+  100+ blocks deep).
+  Similarly, the server MAY ignore new subscription requests if the spending tx is already
+  mined at a reorg-safe height but it still MUST send at least one full response.
```
```suggestion
  mined at a reorg-safe height but it still MUST include all details about the spending transaction in its response to the subscription request .
```

SomberNight replied:
How about:
```
  Similarly, the server MAY ignore new subscription requests if the spending tx is already
  mined at a reorg-safe height but it still MUST send the full response to the subscription request
  (just not the subsequent notifications - which likely would never trigger anyway without a deep reorg).
```

sstone replied:
Yes this is fine.

SomberNight replied:
done in c87b393a2636570d59c75b26f2d7b8cbea921b1e

sstone reviewed (commented):
This is fine by us (we only implement a simple electrum client), I just suggested a rewording of the description of the outpoint subscription (which  will make our life easier :))
 
I understand that version/feature negotiation is a WIP but  if script hash methods are to be removed it could be made stricter (maybe by not using strings to describe protocol versions ?)

SomberNight commented on docs/protocol-basics.rst:
@sstone 

> I understand that version/feature negotiation is a WIP but if script hash methods are to be removed it could be made stricter (maybe by not using strings to describe protocol versions ?)

What do you mean? Do you mean changing the protocol version from string to int, for easier comparison? Or do you mean replacing it with a feature vector?

sstone replied:
Yes, something that is stricter and well defined (changing the version to an int, adding new explicit fields (min/max protocol) with int values, defining explicit enum values that version strings must match, ...).

SomberNight replied:
> adding new explicit fields (min/max protocol)

Note there is already an explicit min/max protocol that the client can send, and from that range the server selects the version to be used. (so the client can send a range, not just a single protocol version)

https://github.com/spesmilo/electrum-protocol/blob/22deba89418c59b5ae5eb5b61cee3e0b33ce3dfb/docs/protocol-methods.rst?plain=1#L1153-L1158
Example:
```
sending message b'{"jsonrpc":"2.0","method":"server.version","id":0,"params":["electrum/4.7.2",["1.4","1.6"]]}'
received data b'{"jsonrpc":"2.0","result":["ElectrumX 1.18.0","1.4.3"],"id":0}\n'
```
One shortcoming however is that the client must support "all" protocol versions in the range it sends :/
(there is some discussion in https://github.com/spesmilo/electrum-protocol/issues/7)
Perhaps in the future the client could send an explicit list of versions in supports - but to make it backwards compatible, it could maybe send this in addition to the min/max range (as long as the client wants to be compatible with older servers).

As for simply changing the version string to a version int, I don't think that would gain us much: it is already fairly simple to compare the string versions. (also, we can't naively change the protocol version strings to ints, as that would be a major breaking change in the version negotiation)

but yes, I see there is a lot of room for improvement...

SomberNight committed:
address feedback: clarify deep reorg for txo sub

ecdsa commented on docs/protocol-basics.rst:
```diff
@@ -224,6 +224,30 @@ those two txs are sorted by txid (`80` comes before `e0`).
+which (naively) for a million blocks might take ~50-100 MB of RAM (or disk space).
+
+To save some bandwidth, we can strip the leading zeroes (from the hex representation!)
+of the block hash. We only strip the zero hex characters from the low-entropy end,
```
This "low-entropy end" is very obscure. If your goal is to disambiguate which end of the hash is being truncated, then it would be better to refer to the mining process explicitly: "To save some bandwidth, we strip the leading zeroes (which result from mining difficulty) from the hex representation of the block hash".

SomberNight replied:
done in 5a702d46d82b554c3d4287d4a57f5a38bfbc797b

SomberNight committed:
(trivial) blockref: rephrase stripping leading zeroes from hash
